content/oci/oci.go

```diff
@@ -0,0 +1,228 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/opencontainers/go-digest"
+	specs "github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/graph"
+	reslv "oras.land/oras-go/v2/internal/resolver"
+)
+
+// ociImageIndexFile is the file name of the index from the OCI Image Layout Specification.
+// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md#indexjson-file
+const ociImageIndexFile = "index.json"
+
+// Store implements `oras.Target`, and represents a content store
+// based on file system with the OCI-Image layout.
+// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md
+type Store struct {
+	storage  content.Storage
+	resolver *reslv.Memory
+	finder   *graph.Memory
+	index    *ocispec.Index
+
+	root      string
+	indexPath string
+}
+
+func New(root string) (*Store, error) {
+	store := &Store{
+		storage:   NewStorage(root),
+		resolver:  reslv.NewMemory(),
+		finder:    graph.NewMemory(),
+		root:      root,
+		indexPath: filepath.Join(root, ociImageIndexFile),
+	}
+
+	if err := store.ensureOCILayoutFile(); err != nil {
+		return nil, err
+	}
+
+	if err := store.loadIndex(); err != nil {
+		return nil, err
+	}
+
+	return store, nil
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (s *Store) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	return s.storage.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (s *Store) Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error {
+	if err := s.storage.Push(ctx, expected, reader); err != nil {
+		return err
+	}
+
+	return s.finder.Index(ctx, s.storage, expected)
+}
+
+// Exists returns true if the described content exists.
+func (s *Store) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	return s.storage.Exists(ctx, target)
+}
+
+// Tag tags a descriptor with a reference string.
+func (s *Store) Tag(ctx context.Context, desc ocispec.Descriptor, reference string) error {
+	if reference == "" {
+		return errdef.ErrMissingReference
+	}
+
+	exists, err := s.storage.Exists(ctx, desc)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("%s: %s: %w", desc.Digest, desc.MediaType, errdef.ErrNotFound)
+	}
+
+	// reference can be either a tag or a digest.
+	if isDigest(reference) {
+		// a digest reference must match the digest of the descriptor
+		if reference != desc.Digest.String() {
+			return fmt.Errorf("digest reference mismatch: %s: expect %s: %w", reference, desc.Digest.String(), errdef.ErrInvalidReference)
+		}
+	} else {
+		if desc.Annotations == nil {
+			desc.Annotations = map[string]string{}
+		}
+		desc.Annotations[ocispec.AnnotationRefName] = reference
```

Comment:
This is a bit tricky. Actually I don't see a need to store the FQDN, we may open an issue to discuss it if needed.
`containerd` stores the FQDN in the `io.containerd.image.name` annotation, but it would be strange if we use the name containing "containerd". We may need to come up with a new name.
Besides, we store the passed reference string in `org.opencontainers.image.ref.name` annotation, which is supposed to be a tag (or digest), how do we know the FQDN of the descriptor?